pkg/crawl/identify.go

```diff
@@ -104,53 +106,98 @@ type Identifier interface {
 // - vulnerable log4j jar files.
 // - zipped files containing vulnerable log4j files, using the provided ZipFileLister.
-func (i *identifier) Identify(ctx context.Context, path string, d fs.DirEntry) (Finding, Versions, error) {
-	if i.listTimeout > 0 {
-		ctxWithTimeout, cancel := context.WithTimeout(ctx, i.listTimeout)
+func (i *Log4jIdentifier) Identify(ctx context.Context, path string, d fs.DirEntry) (Finding, Versions, error) {
+	if i.ArchiveWalkTimeout > 0 {
+		ctxWithTimeout, cancel := context.WithTimeout(ctx, i.ArchiveWalkTimeout)
 		defer cancel()
 		ctx = ctxWithTimeout
 	}
 
 	lowercaseFilename := strings.ToLower(d.Name())
+	versions := make(Versions)
+	result := NothingDetected
+	archiveVersion, match := fileNameMatchesLog4jVersion(lowercaseFilename)
+	if match {
+		if vulnerableVersion(archiveVersion) {
+			result |= JarName
+			versions[archiveVersion] = struct{}{}
+		} else {
+			return result, versions, nil
+		}
+	} else {
+		archiveVersion = UnknownVersion
+	}
 	if hasZipFileEnding(lowercaseFilename) {
-		return i.lookForMatchInZip(ctx, path, lowercaseFilename, UnknownVersion)
+		reader, err := i.OpenFileZipReader(path)
+		if err != nil {
+			return 0, nil, err
+		}
+		defer func() {
+			if cErr := reader.Close(); err == nil && cErr != nil {
+				err = cErr
+			}
+		}()
+		inZip, inZipVs, err := i.lookForMatchInZip(ctx, 0, &reader.Reader, archiveVersion)
+		for v := range inZipVs {
+			versions[v] = struct{}{}
+		}
+		return result | inZip, versions, err
 	}
 	if hasTgzFileEnding(lowercaseFilename) {
 		return i.lookForMatchInTar(ctx, path)
 	}
 	return NothingDetected, nil, nil
 }
 
-func (i *identifier) lookForMatchInZip(ctx context.Context, path string, lowercaseFilename string, parentVersion string) (Finding, Versions, error) {
+func (i *Log4jIdentifier) lookForMatchInZip(ctx context.Context, depth uint, r *zip.Reader, parentVersion string) (Finding, Versions, error) {
+	if depth > i.ArchiveMaxDepth {
+		return 0, nil, nil
+	}
+
 	archiveResult := NothingDetected
 	versions := Versions{}
-	archiveVersion, match := fileNameMatchesLog4jVersion(lowercaseFilename)
-	if !match {
-		archiveVersion = parentVersion
-	} else if vulnerableVersion(archiveVersion) {
-		archiveResult |= JarName
-		versions[archiveVersion] = struct{}{}
-	}
-	err := i.zipWalker(ctx, path, func(ctx context.Context, filename string, size int64, contents io.Reader) (proceed bool, err error) {
-		finding, version, err := lookForMatchInFileInZip(filename, size, contents, archiveVersion)
+	err := i.ZipWalker(ctx, r, func(ctx context.Context, path string, size int64, contents io.Reader) (proceed bool, err error) {
+		if hasZipFileEnding(path) {
+			_, filename := filepath.Split(path)
+			archiveVersion, match := fileNameMatchesLog4jVersion(strings.ToLower(filename))
+			if match {
+				if vulnerableVersion(archiveVersion) {
+					archiveResult |= JarNameInsideArchive
+					versions[archiveVersion] = struct{}{}
+				}
+				return true, nil
```

Comment:
We still want to check inside - it's valuable to know if it's just a name match or the md5 etc also matches from classes inside.